server/src/main/java/keywhiz/service/providers/ClientAuthFactory.java

```diff
@@ -34,58 +44,253 @@
+
+    // Extract client information based on the x-forwarded-client-cert header or
+    // on the security context of this request
+    if (possibleXfccConfig.isEmpty()) {
+      // The XFCC header is not used; use the security context of this request to identify the client
+      return authorizeClient(requestPrincipal, requestName, requestSpiffeId);
+    } else {
+      return authorizeClientFromXfccHeader(possibleXfccConfig.get(), xfccHeaderValues, requestName,
+          requestSpiffeId);
+    }
   }
 
   static Optional<Principal> getPrincipal(ContainerRequest request) {
     return Optional.ofNullable(request.getSecurityContext().getUserPrincipal());
   }
 
-  static Optional<String> getClientName(Optional<Principal> principal) {
-    if (!principal.isPresent()) {
-      return Optional.empty();
-    }
-
-    X500Name name = new X500Name(principal.get().getName());
+  static Optional<String> getClientName(Principal principal) {
+    X500Name name = new X500Name(principal.getName());
     RDN[] rdns = name.getRDNs(BCStyle.CN);
     if (rdns.length == 0) {
-      logger.warn("Certificate does not contain CN=xxx,...: {}", principal.get().getName());
+      logger.warn("Certificate does not contain CN=xxx,...: {}", principal.getName());
       return Optional.empty();
     }
     return Optional.of(IETFUtils.valueToString(rdns[0].getFirst().getValue()));
   }
 
+  private Optional<XfccSourceConfig> getXfccConfigForPort(int port) {
+    List<XfccSourceConfig> matchingConfigs = clientAuthConfig.xfccConfigs()
+        .stream()
+        .filter(xfccSourceConfig -> xfccSourceConfig.port().equals(port))
+        .collect(Collectors.toUnmodifiableList());
+
+    if (matchingConfigs.size() > 1) {
+      throw new NotAuthorizedException(format(
+          "Invalid 'xfcc' configuration for port %d; at most one configuration must be present per port",
+          port));
+    } else {
+      return matchingConfigs.stream().findFirst();
+    }
+  }
+
+  private Client authorizeClientFromXfccHeader(XfccSourceConfig xfccConfig,
+      List<String> xfccHeaderValues, Optional<String> requestName,
+      Optional<String> requestSpiffeId) {
+    // Do not allow the XFCC header to be set by all incoming traffic. This throws a
+    // NotAuthorizedException when the traffic is not coming from a source allowed to set the
+    // header.
+    validateXfccHeaderAllowed(xfccConfig, requestName, requestSpiffeId);
+
+    // Extract client information from the XFCC header
+    Optional<X509Certificate> clientCert =
+        getClientCertFromXfccHeaderEnvoyFormatted(xfccHeaderValues);
+    if (clientCert.isEmpty()) {
+      throw new NotAuthorizedException(
+          format("unable to parse client certificate from %s header", XFCC_HEADER_NAME));
+    }
+
+    Principal clientPrincipal = clientCert.get().getSubjectX500Principal();
+    Optional<String> clientName = getClientName(clientPrincipal);
+    Optional<String> clientSpiffeId = Optional.empty();
```

Comment:
It's not immediately clear that SPIFFE IDs are not supported, yet (even though the authentication logic already exists in some places). Additional comment here would be great!